substrate/frame/election-provider-support/src/lib.rs

```diff
@@ -361,72 +379,87 @@ pub trait ElectionDataProvider {
 	/// Clear all voters and targets.
 	#[cfg(any(feature = "runtime-benchmarks", test))]
 	fn clear() {}
+
+	#[cfg(any(feature = "runtime-benchmarks", test))]
+	fn set_desired_targets(_count: u32) {}
 }
 
-/// Base trait for types that can provide election
-pub trait ElectionProviderBase {
-	/// The account identifier type.
+/// An [`ElectionDataProvider`] that exposes for an external entity to request a lock/unlock on
+/// updates in the election data.
+pub trait LockableElectionDataProvider: ElectionDataProvider {
+	fn set_lock() -> data_provider::Result<()>;
+	fn unlock();
+}
+
+/// Something that can compute the result of an election and pass it back to the caller in a paged
+/// way.
+pub trait ElectionProvider {
+	/// The account ID identifier;
 	type AccountId;
 
 	/// The block number type.
 	type BlockNumber;
 
-	/// The error type that is returned by the provider.
-	type Error: Debug;
+	/// The error type returned by the provider;
+	type Error: Debug + PartialEq;
 
-	/// The upper bound on election winners that can be returned.
-	///
-	/// # WARNING
-	///
-	/// when communicating with the data provider, one must ensure that
-	/// `DataProvider::desired_targets` returns a value less than this bound. An
-	/// implementation can chose to either return an error and/or sort and
-	/// truncate the output to meet this bound.
-	type MaxWinners: Get<u32>;
+	/// The maximum number of winners per page in results returned by this election provider.
+	type MaxWinnersPerPage: Get<u32>;
+
+	/// The maximum number of backers that a single page may have in results returned by this
+	/// election provider.
+	type MaxBackersPerWinner: Get<u32>;
+
+	/// The number of pages that this election provider supports.
+	type Pages: Get<PageIndex>;
 
 	/// The data provider of the election.
 	type DataProvider: ElectionDataProvider<
 		AccountId = Self::AccountId,
 		BlockNumber = Self::BlockNumber,
 	>;
 
+	/// Elect a new set of winners.
+	///
+	/// The result is returned in a target major format, namely as vector of supports.
+	///
+	/// This should be implemented as a self-weighing function. The implementor should register its
+	/// appropriate weight at the end of execution with the system pallet directly.
+	fn elect(remaining: PageIndex) -> Result<BoundedSupportsOf<Self>, Self::Error>;
+
+	/// The index of the *most* significant page that this election provider supports.
+	fn msp() -> PageIndex {
```

Comment:
may be call it highest page or top page?

Reply:
Also, since the current page index needs to be stored somewhere, could we just store it on `ElectionProvider`, and Staking then can just call `next_elect()`?

Reply:
I think I called it like this initially be be simiar to the concept if msp (most significant bit) in cpus.